### DISCUSSION TITLE
Using Shortcodes with React lazy

### DISCUSSION THREAD

**Discussion by tnorlund** [mdx-js/mdx#1492] [Q&A] [answered]
I'd like to split up my components using React lazy. Can I do this using MDX shortcodes? I'm currently getting this error:

> Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object. Check the render method of `MDXCreateElement`.

```js
...
const CICD = React.lazy( () => import( 'components/Icons/CICD' ) )
...
const components = { 
  DocsGrid,
  CICD: <>
    {
      typeof window ==! "undefined" && <React.Suspense 
        fallback={<div height={`200pt`}></div>}
      ><CICD /></React.Suspense>
    }
  </>
}
...
<MDXProvider components={components}>
    {children}
```

**Comment by ChristianMurphy** [comment, marked as answer]
The error you're getting is because you are passing a rendered element instead of a component.

```js
const components = { 
  DocsGrid,
  CICD: () => (<>
    {
      typeof window ==! "undefined" && <React.Suspense 
        fallback={<div height={`200pt`}></div>}
      ><CICD /></React.Suspense>
    }
  </>)
}
```

should work.

**Reply by tnorlund** [reply]
I get the same thing with this:
```js
const components = { 
  DocsGrid,
  CICD: () => ( <>
    {
      typeof window ==! "undefined" && <React.Suspense 
        fallback={<div height={`200pt`}></div>}
      ><CICD /></React.Suspense>
    }
  </> )
}
```

**Reply by ChristianMurphy** [reply]
could you share a code sandbox with a runnable example of the issue?
Here's an example of how it can work
https://codesandbox.io/s/hardcore-browser-4uvt1

**Reply by tnorlund** [reply]
It's there. Thanks for the help!

**Reply by tnorlund** [reply]
Maybe the SSR check is breaking it? It seems to work fine when not constructed into the `components` Object.

**Reply by tnorlund** [reply]
It won't build without the `typeof window ==! "undefined" &&` due to SSR. When I build with that, the page only shows the fallback component. Is there a way to lazy load components in MDX?

**Reply by JounQin** [reply]
Maybe  you can try `typeof window ==! "undefined" ? <></> : null`

**Reply by tnorlund** [reply]
> Maybe you can try `typeof window ==! "undefined" ? <></> : null`

This causes the SSR error.

Outside of MDX, I needed to use the condition within the component: 
```js
return <>
{ !isSSR <React.Suspense fallback={<div height={`200pt`}></div>}><CICD /></React.Suspense>}
</>
```

**Reply by tnorlund** [reply]
So I just tried these 2 examples.

```js
// Example A
const components = { 
  DocsGrid,
  CICD: () => ( <div style={{ height: `200pt` }} >Loading...</div> )
}
```

```js
// Example B
const components = { 
  DocsGrid,
  CICD: () => ( <>{typeof window ==! "undefined" && <div style={{ height: `200pt` }} >Loading...</div>}</> )
}
```

Example A rendered the div fine. After building with Example B, there's no div and the component is skipped entirely.

**Reply by tnorlund** [reply]
The latest ReactDOM doesn't support react suspense; this is why you need the `typeof window ==! "undefined" && ...`. I found a package for splitting code that supports SSR and ReactDOM. I wish there was a vanilla answer to this, but it looks like it just isn't supported yet.

```js
import loadable from '@loadable/component'

const components = { 
  DocsGrid,
  Analytics: loadable( () => import( 'components/Flow/Analytics' ) ),
...
```
